src/itemmenu.cpp

```diff
@@ -952,15 +808,110 @@ static void sort_items()
  */
 int useup_item(int item_id)
 {
-    size_t inventory_index;
+    return g_inv.remove(item_id) ? 1 : 0;
+}
+
+// KInventory implementation
+
+s_inventory KInventory::operator[](int i)
+{
+    if (i >= 0 && i < inv.size())
+    {
+        return inv.at(i);
+    }
+    else
+    {
+        return {};
+    }
+}
 
-    for (inventory_index = 0; inventory_index < MAX_INV; ++inventory_index)
+void KInventory::add(int type, int quantity)
+{
+    inv.emplace_back(type, quantity);
+    normalize();
+}
+bool KInventory::remove(int type, int quantity)
+{
+    for (auto& it : inv)
     {
-        if (g_inv[inventory_index].item == item_id)
+        if (it.item == type)
         {
-            remove_item(inventory_index, 1);
-            return 1;
+            if (it.quantity >= quantity)
+            {
+                it.quantity -= quantity;
+                quantity = 0;
+            }
+            else
+            {
+                quantity -= it.quantity;
```

Comment:
If we have 9 in one slot and 1 more in another (10 total), and `remove(I_XXX, 10)`, will this leave a gap in the inventory?

Reply:
The inventory gets 'normalised' after being modified which closes up any gaps.